[PREVIEW] New tool ability GUI crashes the server

When switching tool ability, it crashes the server
---- Minecraft Crash Report ----
// You're mean.

Time: 26.05.25 15:05
Description: Exception in server tick loop

java.lang.NoClassDefFoundError: net/minecraft/client/resources/I18n
	at com.hbm.handler.ability.IBaseAbility.getFullName(IBaseAbility.java:13)
	at com.hbm.handler.ability.ToolPreset.getMessage(ToolPreset.java:34)
	at com.hbm.items.tool.ItemToolAbility.func_77659_a(ItemToolAbility.java:280)
	at net.minecraft.item.ItemStack.func_77957_a(ItemStack.java:146)
	at net.minecraft.server.management.ItemInWorldManager.func_73085_a(ItemInWorldManager.java:316)
	at net.minecraft.network.NetHandlerPlayServer.func_147346_a(NetHandlerPlayServer.java:539)
	at net.minecraft.network.play.client.C08PacketPlayerBlockPlacement.func_148833_a(SourceFile:60)
	at net.minecraft.network.play.client.C08PacketPlayerBlockPlacement.func_148833_a(SourceFile:9)
	at net.minecraft.network.NetworkManager.func_74428_b(NetworkManager.java:212)
	at net.minecraft.network.NetworkSystem.func_151269_c(NetworkSystem.java:165)
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:659)
	at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:334)
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:547)
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:396)
	at net.minecraft.server.MinecraftServer$2.run(MinecraftServer.java:685)
Caused by: java.lang.ClassNotFoundException: net.minecraft.client.resources.I18n
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:191)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 15 more
Caused by: java.lang.NullPointerException
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:182)
	... 17 more

--- a/src/main/java/com/hbm/handler/ability/IBaseAbility.java
+++ b/src/main/java/com/hbm/handler/ability/IBaseAbility.java
@@ -1,6 +1,6 @@
 package com.hbm.handler.ability;
 
-import com.hbm.util.i18n.I18nUtil;
+import net.minecraft.client.resources.I18n;
 
 public interface IBaseAbility extends Comparable<IBaseAbility> {
 	public String getName();
@@ -9,9 +9,9 @@ public interface IBaseAbility extends Comparable<IBaseAbility> {
 		return "";
 	}
 
+	// Note: only usable client-side. Server-side, use ChatComponentTranslation manually instead
 	public default String getFullName(int level) {
-		//bandaid fix so it doesn't crash servers instantly, TODO: use ChatComponentTranslation
-		return I18nUtil.format(getName()) + getExtension(level);
+		return I18n.format(getName()) + getExtension(level);
 	}
 
 	public default boolean isAllowed() {

--- a/src/main/java/com/hbm/handler/ability/ToolPreset.java
+++ b/src/main/java/com/hbm/handler/ability/ToolPreset.java
@@ -27,26 +27,32 @@ public class ToolPreset {
 		this.harvestAbilityLevel = harvestAbilityLevel;
 	}
 
-	public ChatComponentText getMessage() {
-		if(isNone())
-			return ChatBuilder.start("[Tool ability deactivated]").color(EnumChatFormatting.GOLD).flush();
+    public ChatComponentText getMessage() {
+        if (isNone()) {
+            return ChatBuilder.start("[Tool ability deactivated]").color(EnumChatFormatting.GOLD).flush();
+        }
 
-		String areaPart = areaAbility.getFullName(areaAbilityLevel);
-		String harvestPart = harvestAbility.getFullName(harvestAbilityLevel);
+        boolean hasArea = areaAbility != IToolAreaAbility.NONE;
+        boolean hasHarvest = harvestAbility != IToolHarvestAbility.NONE;
+        
+        ChatBuilder builder = ChatBuilder.start("[Enabled ");
 
-		ChatBuilder builder = ChatBuilder.start("[Enabled ");
+        if (hasArea) {
+            builder.nextTranslation(areaAbility.getName());
+            builder.next(areaAbility.getExtension(areaAbilityLevel));
+        }
 
-		if(!areaPart.isEmpty())
-			builder.next(areaPart);
+        if (hasArea && hasHarvest) {
+            builder.next(" + ");
+        }
 
-		if(!areaPart.isEmpty() && !harvestPart.isEmpty())
-			builder.next(" + ");
-
-		if(!harvestPart.isEmpty())
-			builder.next(harvestPart);
-
-		return builder.colorAll(EnumChatFormatting.YELLOW).flush();
-	}
+        if (hasHarvest) {
+            builder.nextTranslation(harvestAbility.getName());
+            builder.next(harvestAbility.getExtension(harvestAbilityLevel));
+        }
+        
+        return builder.colorAll(EnumChatFormatting.YELLOW).flush();
+    }
 
 	public boolean isNone() {
 		return areaAbility == IToolAreaAbility.NONE && harvestAbility == IToolHarvestAbility.NONE;

